src/drivers/magnetometer/bosch/bmm350/BMM350.cpp

```diff
@@ -0,0 +1,683 @@
+				_state = STATE::RESET;
+				ScheduleDelayed(30_ms);
+			}
+		}
+		break;
+
+	case STATE::AFTER_RESET:
+		uint8_t chipId;
+
+		if (((chipId = RegisterRead(Register::CHIP_ID)) == chip_identification_number)) {
+
+			// Prep self test
+			RegisterWrite(Register::PMU_CMD, PMU_CMD_SUSPEND);
+			px4_usleep(30000);
+			uint8_t odr_reg_data = (ODR_100HZ & 0xf);
+			odr_reg_data = ((odr_reg_data & ~(0x30)) | ((AVG_2 << 0x4) & 0x30));
+			RegisterWrite(Register::PMU_CMD_AGGR_SET, odr_reg_data);
+			RegisterWrite(Register::PMU_CMD_AXIS_EN, EN_XYZ);
+			RegisterWrite(Register::PMU_CMD, PMU_CMD_FGR);
+			px4_usleep(30000);
+			RegisterWrite(Register::PMU_CMD, PMU_CMD_BR_FAST);
+			px4_usleep(4000);
+			RegisterWrite(Register::PMU_CMD, PMU_CMD_FAST_FM);
+
+			PX4_DEBUG("Chip id found going to self test id= %i", chipId);
+			_state = STATE::SELF_TEST_CHECK;
+			ScheduleDelayed(10_ms);
+
+		} else {
+			if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				PX4_DEBUG("Reset failed, retrying");
+				_state = STATE::RESET;
+				ScheduleDelayed(100_ms);
+
+			} else {
+				PX4_DEBUG("Reset not complete, check again in 10 ms");
+				ScheduleDelayed(10_ms);
+			}
+		}
+
+		break;
+
+	case STATE::SELF_TEST_CHECK: {
+
+			float out_ust[4] = {0.0f};
+
+			float out_ustxh = 0.0f, out_ustxl = 0.0f, out_ustyh = 0.0f, out_ustyl = 0.0f, out_ust_x = 0.0f, out_ust_y = 0.0f;
+			RegisterWrite(Register::TMR_SELF_TEST_USER, SELF_TEST_POS_X);
+			px4_usleep(1000);
+			RegisterWrite(Register::PMU_CMD, PMU_CMD_FAST_FM);
+			px4_usleep(6000);
+			ReadOutRawData(out_ust);
+			out_ustxh = out_ust[0];
+
+			RegisterWrite(Register::TMR_SELF_TEST_USER, SELF_TEST_NEG_X);
+			px4_usleep(1000);
+			RegisterWrite(Register::PMU_CMD, PMU_CMD_FAST_FM);
+			px4_usleep(6000);
+			ReadOutRawData(out_ust);
+			out_ustxl = out_ust[0];
+
+
+			RegisterWrite(Register::TMR_SELF_TEST_USER, SELF_TEST_POS_Y);
+			px4_usleep(1000);
+			RegisterWrite(Register::PMU_CMD, PMU_CMD_FAST_FM);
+			px4_usleep(6000);
+			ReadOutRawData(out_ust);
+			out_ustyh = out_ust[1];
+
+			RegisterWrite(Register::TMR_SELF_TEST_USER, SELF_TEST_NEG_Y);
+			px4_usleep(1000);
+			RegisterWrite(Register::PMU_CMD, PMU_CMD_FAST_FM);
+			px4_usleep(6000);
+			ReadOutRawData(out_ust);
+			out_ustyl = out_ust[1];
+
+			out_ust_x = out_ustxh - out_ustxl;
+			out_ust_y = out_ustyh - out_ustyl;
+
+			PX4_DEBUG("outustxh = %.5f, outustxl = %.5f", static_cast<double>(out_ustxh), static_cast<double>(out_ustxl));
+			PX4_DEBUG("outustyh = %.5f, outustyl = %.5f", static_cast<double>(out_ustyh), static_cast<double>(out_ustyl));
+			PX4_DEBUG("out_ust_x = %.5f, out_ust_y = %.5f", static_cast<double>(out_ust_x), static_cast<double>(out_ust_y));
+
+			// Datasheet 5.1.6
+			if (out_ust_x >= 130 && out_ust_y >= 130) {
+				PX4_DEBUG("Running to configure");
+				_state = STATE::CONFIGURE;
+				ScheduleDelayed(10_ms);
+
+			} else if (perf_event_count(_self_test_failed_perf) >= 5) {
+				PX4_DEBUG("Failed after 5 attempts, procceed still");
```

Comment:
Why proceed even if self test fails? I believe one should try to recover via reset instead.